### PR TITLE
[BOLT] Deduplicate equal offsets in BAT

### PR DESCRIPTION
**Pull request by aaupov**
Encode BRANCHENTRY bits as bitmask for deduplicated entries.

Reduces BAT section size:
- large binary: to 11834216 bytes (0.31x original),
- medium binary: to 1565584 bytes (0.26x original),
- small binary: to 336 bytes (0.23x original).

Test Plan: Updated bolt/test/X86/bolt-address-translation.test

### CODE REVIEW

File: bolt/lib/Profile/BoltAddressTranslation.cpp [hunk 1]
```diff
@@ -207,15 +280,23 @@ void BoltAddressTranslation::parseMaps(std::vector<uint64_t> &HotFuncs,
...
           dbgs() << formatv("{0:x} -> {1:x} ({2}/{3}b -> {4}/{5}b), {6:x}\n",
                             OutputOffset, InputOffset, OutputDelta,
-                            encodeULEB128(OutputDelta, nulls()), InputDelta,
-                            encodeSLEB128(InputDelta, nulls()), OutputAddress));
+                            getULEB128Size(OutputDelta), InputDelta,
+                            (J < EqualElems) ? 0 : getSLEB128Size(InputDelta),
+                            OutputAddress));
     }
+    if (BEBitMask)
+      delete BEBitMask;
```
Comment (dcci) [inline bolt/lib/Profile/BoltAddressTranslation.cpp:299]: Can you make this automatically memory managed?

File: bolt/lib/Profile/BoltAddressTranslation.cpp [hunk 2]
```diff
@@ -110,6 +111,34 @@ void BoltAddressTranslation::write(const BinaryContext &BC, raw_ostream &OS) {
...
+
+size_t BoltAddressTranslation::getNumEqualOffsets(const MapTy &Map) const {
+  size_t EqualOffsets = 0;
+  for (const std::pair<const uint32_t, uint32_t> &KeyVal : Map) {
+    const uint32_t OutputOffset = KeyVal.first;
+    const uint32_t InputOffset = KeyVal.second >> 1;
+    if (OutputOffset == InputOffset)
+      ++EqualOffsets;
+    else
+      break;
```
Comment (dcci) [inline bolt/lib/Profile/BoltAddressTranslation.cpp:137]: Do you want to stop here at the first offset that's different?

Comment (aaupov) [reply bolt/lib/Profile/BoltAddressTranslation.cpp:137]: Yes, that's by design. The idea is that if OutputOffset==InputOffset, only emit one of them. It's expected that if offsets start to diverge at some point, they're not reconverging.

File: bolt/lib/Profile/BoltAddressTranslation.cpp (hunk 1 above)
Comment (aaupov) [reply bolt/lib/Profile/BoltAddressTranslation.cpp:299]: Good point. I can keep APInt uninitialized if it's not used.

**Review by dcci — APPROVED** [review]
LG

**Review by rafaelauler — APPROVED** [review]

**Pull request merged**